src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs

```diff
@@ -0,0 +1,371 @@
+    if !binary_path.exists() {
+        return Err(anyhow!("Hook Binary path {:?} does not exist", binary_path));
+    }
+
+    // hook.args[0] is the binary name, which should not be included in Command.args
+    let mut command_args = h.args.clone();
+    if !command_args.is_empty() {
+        command_args.remove(0);
+    }
+
+    // split envs, and those invalid envs will be omitted
+    let command_envs: HashMap<&str, &str> = h.env.iter().filter_map(|e| valid_env(e)).collect();
+
+    // set execution timeout duration
+    let execution_timeout = match h.timeout {
+        Some(t) => Duration::new(t as u64, 0),
+        None => Duration::MAX,
+    };
+
+    // create tokio child process to execute hook
+    let mut child = tokio::process::Command::new(&binary_path)
+        .args(command_args.iter())
+        .envs(command_envs.iter())
+        .kill_on_drop(true)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let state = serde_json::to_string(st)?;
+    let join_handle = tokio::spawn(async move {
+        // write state to stdin of child process
+        if let Some(mut stdin) = child.stdin.take() {
+            match stdin.write_all(state.as_bytes()).await {
+                Ok(_) => {}
+                Err(e) => {
+                    info!(sl!(), "write to child stdin error: {:?}", e);
+                }
+            }
+        }
+
+        // read info from stdout of child process
+        if let Some(stdout) = child.stdout.as_mut() {
+            let mut out = String::new();
+            match stdout.read_to_string(&mut out).await {
+                Ok(_) => {
+                    info!(sl!(), "child stdout: {:?}", out.as_str());
+                }
+                Err(e) => {
+                    info!(sl!(), "read from child stdout failed: {:?}", e);
+                }
+            }
+        }
+
+        // read err info from stderr of child process
+        let mut err = String::new();
+        if let Some(stderr) = child.stderr.as_mut() {
+            match stderr.read_to_string(&mut err).await {
+                Ok(_) => {
+                    info!(sl!(), "child stderr: {:?}", err.as_str());
+                }
+                Err(e) => {
+                    info!(sl!(), "read from child stderr failed: {:?}", e);
+                }
+            }
+        }
+
+        match child.wait().await {
+            Ok(exit_status) => {
+                let exit_code = exit_status
+                    .code()
+                    .ok_or_else(|| anyhow!("hook exit status has no code"))?;
+                if exit_code != 0 {
+                    error!(
+                        sl!(),
+                        "hook {} exit code: {}, error message: {}",
+                        &binary_path.display(),
+                        exit_code,
+                        err
+                    );
+                    return Err(anyhow!(nix::Error::UnknownErrno));
```

Comment:
Is it better to return more detailed error like lines 152-156?

Reply:
The problem here is that there may be many kinds of errors, and it is difficult for us to determine which one.

Reply:
The exitcode should be a Kernel Errno, you can convert that to Error. I think libc or nix can do that.

Reply:
I'm not sure whether it's a Kernel Errno at all here, and I need sometime to confirm ...